platform/xdebugger-impl/src/com/intellij/xdebugger/impl/XDebuggerActionsCollector.kt

```diff
@@ -0,0 +1,90 @@
+import com.intellij.internal.statistic.eventLog.events.EventFields.Int
+import com.intellij.internal.statistic.eventLog.events.EventFields.IntList
+import com.intellij.internal.statistic.eventLog.events.EventFields.String
+import com.intellij.internal.statistic.eventLog.events.EventFields.StringListValidatedByCustomRule
+import com.intellij.internal.statistic.eventLog.validator.ValidationResultType.ACCEPTED
+import com.intellij.internal.statistic.eventLog.validator.rules.EventContext
+import com.intellij.internal.statistic.eventLog.validator.rules.impl.CustomValidationRule
+import com.intellij.internal.statistic.service.fus.collectors.CounterUsagesCollector
+import com.intellij.xdebugger.frame.XStackFrame
+
+private const val UNKNOWN_TYPE = "Unknown"
+
+class XDebuggerActionsCollector : CounterUsagesCollector() {
+  companion object {
+
+    private val group = EventLogGroup("xdebugger.actions", 3)
+    const val PLACE_THREADS_VIEW = "threadsView"
+    const val PLACE_FRAMES_VIEW = "framesView"
+    private const val LOCATION = "location"
+    @Suppress("MemberVisibilityCanBePrivate") // Exposed for consumers of this metric
+    const val EVENT_FRAMES_UPDATED = "frames.updated"
+    @Suppress("MemberVisibilityCanBePrivate") // Exposed for consumers of this metric
+    const val TOTAL_FRAMES = "total_frames"
+    @Suppress("MemberVisibilityCanBePrivate") // Exposed for consumers of this metric
+    const val FILE_TYPES = "file_type"
+    @Suppress("MemberVisibilityCanBePrivate") // Exposed for consumers of this metric
+    const val FRAMES_PER_TYPE = "frames_per_file_type"
+    private val durationField = EventFields.DurationMs
+    private val totalFramesField = Int(TOTAL_FRAMES)
+    private val frameTypesField = StringListValidatedByCustomRule<FrameTypeValidator>(FILE_TYPES)
+    private val framesPerTypesField = IntList(FRAMES_PER_TYPE)
+    private val framesUpdated =
+      group.registerVarargEvent(EVENT_FRAMES_UPDATED, durationField, totalFramesField, frameTypesField, framesPerTypesField)
+
+    @JvmField
+    val threadSelected = group.registerEvent("thread.selected", String(LOCATION, listOf(PLACE_FRAMES_VIEW, PLACE_THREADS_VIEW)))
+
+    @JvmField
+    val frameSelected = group.registerEvent("frame.selected", String(LOCATION, listOf(PLACE_FRAMES_VIEW, PLACE_THREADS_VIEW)))
+
+    @JvmField
+    val sessionChanged = group.registerEvent("session.selected")
+
+    @JvmStatic
+    fun logFramesUpdated(durationMs: Long, frames: List<XStackFrame>) {
+      val framesByType = frames.groupingBy { it.getFrameType() }.eachCount().entries
+      val fileTypes = framesByType.map { it.key }
+      val counts = framesByType.map { it.value }
+
+      framesUpdated.log(
+        durationField.with(durationMs),
+        totalFramesField.with(frames.size),
+        frameTypesField.with(fileTypes),
+        framesPerTypesField.with(counts)
+      )
+    }
+  }
+
+  private class FrameTypeValidator : CustomValidationRule() {
```

Comment:
You need to register each custom validation rule in xml as EP.
```xml
    <statistics.validation.customValidationRule
      implementation="com.intellij.xdebugger.impl.XDebuggerActionsCollector$FrameTypeValidator"/>
```

And implement `getRuleId()` method with unique id for rule.
```kotlin
override fun getRuleId() = "frame_type"
```

Reply:
Done